### PR TITLE
add missing RBAC to statefulset-controller for StatefulSetAutoDeletePVC feature

### LINKED ISSUE #126919: StatefulSet PersistentVolumeClaimRetentionPolicy not deleting pods when `WhenScaled: Delete` is used
### What happened?

When OwnerReferencesPermissionEnforcement admission plugin is used the PVCs are not deleted and KCM logs show

```
stateful_set.go:438] "Unhandled Error" err="error syncing StatefulSet e2e-statefulset-1256/ss, requeuing: could not update claim datadir-ss-2 for delete policy ownerRefs: persistentvolumeclaims \"datadir-ss-2\" is forbidden: cannot set blockOwnerDeletion if an ownerReference refers to a resource you can't set finalizers on: , <nil>" logger="UnhandledError"
```

### What did you expect to happen?

PVCs should have ownerReferences set and the PVCs should be garbage collected on scale down

### How can we reproduce it (as minimally and precisely as possible)?

3 of 6 of the following tests fail when OwnerReferencesPermissionEnforcement admission plugin is turned on
```
./hack/ginkgo-e2e.sh --ginkgo.focus=".*Non-retain StatefulSetPersistentVolumeClaimPolicy should.*" --ginkgo.v
```

### Anything else we need to know?

_No response_

### Kubernetes version

<details>

```console
$ kubectl version
Client Version: v1.29.8
Kustomize Version: v5.0.4-0.20230601165947-6ce0bf390ce3
Server Version: v1.32.0-alpha.0.233+b3c725627b15bb-dirty

```

</details>

### Cloud provider

<details>
NA
</details>

### OS version

<details>

```console
# On Linux:
$ cat /etc/os-release
# paste output here
$ uname -a
# paste output here

# On Windows:
C:\> wmic os get Caption, Version, BuildNumber, OSArchitecture
# paste output here
```

</details>

### Install tools

<details>

</details>

### Container runtime (CRI) and version (if applicable)

<details>

</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

### PR DESCRIPTION
**Pull request by atiratree**
#### What type of PR is this?

/kind bug
/kind failing-test

#### What this PR does / why we need it:

statefulset-controller RBAC needs do not reflect the reality. See the linked issue for more details.

#### Which issue(s) this PR fixes:

Fixes #126919

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: plugin/pkg/auth/authorizer/rbac/bootstrappolicy/controller_policy.go [hunk 1]
```diff
@@ -383,13 +383,14 @@ func buildControllerRoles() ([]rbacv1.ClusterRole, []rbacv1.ClusterRoleBinding)
 				rbacv1helpers.NewRule("update").Groups(appsGroup).Resources("statefulsets/finalizers").RuleOrDie(),
 				rbacv1helpers.NewRule("get", "create", "delete", "update", "patch").Groups(legacyGroup).Resources("pods").RuleOrDie(),
 				rbacv1helpers.NewRule("get", "create", "delete", "update", "patch", "list", "watch").Groups(appsGroup).Resources("controllerrevisions").RuleOrDie(),
-				rbacv1helpers.NewRule("get", "create").Groups(legacyGroup).Resources("persistentvolumeclaims").RuleOrDie(),
+				rbacv1helpers.NewRule("get", "create", "list", "watch").Groups(legacyGroup).Resources("persistentvolumeclaims").RuleOrDie(),
```
Comment (atiratree) [inline plugin/pkg/auth/authorizer/rbac/bootstrappolicy/controller_policy.go:386]: used by

https://github.com/kubernetes/kubernetes/blob/0bcbc3b77a0b5358f53e81e8fa7605909bf8a2fe/pkg/controller/statefulset/stateful_set.go#L100

File: plugin/pkg/auth/authorizer/rbac/bootstrappolicy/controller_policy.go [hunk 2]
```diff
@@ -383,13 +383,14 @@ func buildControllerRoles() ([]rbacv1.ClusterRole, []rbacv1.ClusterRoleBinding)
...
 				eventsRule(),
 			},
 		}
 
 		if utilfeature.DefaultFeatureGate.Enabled(features.StatefulSetAutoDeletePVC) {
 			role.Rules = append(role.Rules, rbacv1helpers.NewRule("update", "delete").Groups(legacyGroup).Resources("persistentvolumeclaims").RuleOrDie())
+			role.Rules = append(role.Rules, rbacv1helpers.NewRule("update").Groups(legacyGroup).Resources("pods/finalizers").RuleOrDie())
```
Comment (atiratree) [inline plugin/pkg/auth/authorizer/rbac/bootstrappolicy/controller_policy.go:393]: used by
https://github.com/kubernetes/kubernetes/blob/0bcbc3b77a0b5358f53e81e8fa7605909bf8a2fe/pkg/controller/statefulset/stateful_set_utils.go#L338

**Review by soltysh — APPROVED** [review]
/lgtm

**Comment by atiratree** [comment]
unrelated
/retest

**Comment by atiratree** [comment]
I have opened backports for 1.31-1.28, linked above^

**Pull request merged**